synapse/storage/databases/main/events.py

```diff
@@ -826,14 +825,28 @@ def _allocate_chain_ids(
         )
         txn.execute(sql % (clause,), args)
 
-        chain_to_max_seq_no = {row[0]: row[1] for row in txn}  # type: Dict[int, int]
+        chain_to_max_seq_no = {row[0]: row[1] for row in txn}  # type: Dict[Any, int]
 
         # Allocate the new events chain ID/sequence numbers.
-        for event_id, prev_event_id in tree:
-            if prev_event_id:
-                existing_chain_id = chain_map[prev_event_id]
-
-            new_chain_tuple = None
+        #
+        # To reduce the number of calls to the database we don't allocate a
+        # chain ID number in the loop, instead we use a temporary `object()` for
+        # each new chain ID. Once we've done the loop we generate the necessary
+        # number of new chain IDs in one call, replacing all temporary
+        # objects with real allocated chain IDs.
+
+        unallocated_chain_ids = set()  # type: Set[object]
+        new_chain_tuples = {}  # type: Dict[str, Tuple[Any, int]]
+        for event_id, auth_event_id in tree:
+            # If we reference an auth_event_id we fetch the allocated chain ID,
+            # either from the existing `chain_map` or the newly generated
+            # `new_chain_tuples` map.
+            if auth_event_id:
```

Comment:
If `auth_event_id` is not set, I *think* this will use a previous `existing_chain_id`, which I think leaks between loops?

Reply:
oh, woops